### PR TITLE
HPCC-28868 Recode CHttpRequest::parseQueryString to avoid a temporary…

### PR DESCRIPTION
**Pull request by ghalliday**
… buffer

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
@afishbeck I have not tested this change.  How speed critical is this piece of code?  I made the minimal changes to remove the strcpy, but more could be done if performance is important - e.g., reuse the stringbuffers and recode the loop slightly.

File: system/jlib/jstring.cpp [hunk 1]
```diff
@@ -1595,6 +1595,28 @@ StringBuffer &appendDecodedURL(StringBuffer &s, const char *url)
     return s;
 }
 
+StringBuffer &appendDecodedURL(StringBuffer &s, size_t len, const char *url)
+{
+    size_t i = 0;
```
Comment (richardkchapman) [inline system/jlib/jstring.cpp:1600]: Should this function do something different if it encounters a null char?

Comment (ghalliday) [reply system/jlib/jstring.cpp:1600]: If anything it should throw an exception - I'm not sure how a url encoded string can include a null.  I think it is probably enough to pass it through - it doesn't validate any other characters are valid.

**Comment by afishbeck** [comment]
> @afishbeck I have not tested this change. How speed critical is this piece of code? I made the minimal changes to remove the strcpy, but more could be done if performance is important - e.g., reuse the stringbuffers and recode the loop slightly.

This code is called whenever an HTTP request has parameters after the '?' and for an HTTP POST with form encoded content (submitting a web form, or some REST calls).

Most calls that happen in volume would be SOAP or JSON and would have short strings after the '?' if any.

EclWatch does make rest calls in the FORM POST format.. but those usually happen because of user action, rather than large scale.

That doesn't stop someone from making any ESP call using a queryString or FORM POST which are forms of REST calls.  It's just less likely to be used at scale than SOAP or JSON.

That being said, optimizing this further would be great, just not sure of the priority based on what I said above.

**Review by afishbeck — COMMENTED** [review]
@ghalliday your change looks good barring further optimization.  I was going to test but then got trapped in build error hell and I'll need to climb out of that tomorrow.  If you'd like I can test once I'm able.

**Comment by ghalliday** [comment]
@afishbeck thanks.  Testing would be good - I'm in no rush to merge it.

**Review by afishbeck — APPROVED** [review]
@ghalliday I've tested this commit on a different branch to avoid my build issues.  It worked fine and the change looks good.  There is plenty of room for optimizations and commoning up code, but not needed for this PR.

**Pull request merged**